Introduce estimate of the change in junction angle per mm traveled

XDA-Bam opened the pull request:
### Description

_**NOTE:** Re-submitted without duplicate upstream commits._

This introduces `d_theta_indicator` as an exponentially decaying estimate of the change in junction angle per mm traveled. Based on this indicator and a threshold value, the speed limit `limit_sqr` can be kept constant during tiny changes in the junction angle, avoiding stutter and leading to a smoother print. Details are given in posts [[1]](https://github.com/MarlinFirmware/Marlin/issues/17342#issuecomment-617469776) and [[2]](https://github.com/MarlinFirmware/Marlin/issues/17342#issuecomment-617473757).
Essentially, this accumulates changes in `junction_theta` over the traveled path, while at the same time "forgetting" a fixed fraction of the accumulated change per length traveled. If the accumulated value `d_theta_indicator` is below `d_theta_threshold`, we act as if there was no change in angle at all and keep our speed limit `limit_sqr` reasonably high. If `d_theta_indicator` surpasses `d_theta_threshold`, we instantly reset `limit_sqr` to the default value for the current block/segment.

**Notes:** 
- `d_theta_decay` and `d_theta_threshold` should probably be made configurable, if this method is introduced to Marlin.
- Typical values for `d_theta_decay`: `0.3f` (slow decay, angle changes have long lasting effect which leads to less coasting) to `1.0f` (fast decay, coasting is seen more often).
- Typical values for `d_theta_threshold`: `RADIANS(0.001f)` (very sensitive, little coasting) to `RADIANS(0.5f)` (very low sensitivity, coasting is often applied and many angle changes do not affect `limit_sqr`)

### Benefits

This reduces frequent changes in `limit_sqr` during printing of nearly straight segments of a part. It also avoids some dips in `limit_sqr` caused by tiny segments combined with an unfortunate distribution of angles around those segments as found in #17342 .

### Not-so-Benefits

Currently requires `limit_sqr` and `junction_theta` to be calculated for every move, independent of segment length and junction angle. This uses additional resources. Feel free to suggest any solution where this can be avoided.

### Related Issues

Further mitigates #17342 .

XDA-Bam committed:
Introduce estimate of the change in junction angle per mm traveled

This introduces del_angle_indicator as an exponentially decaying estimate of the change in junction angle per mm traveled. Based on this indicator and a threshold value, the speed limit limit_sqr can be kept constant during tiny changes in the junction angle, avoiding stutter and leading to a smoother print. This further mitigates #17342 .

XDA-Bam committed:
Rename junction angle change indicator

del_angle_xyz -> d_theta_xyz
Just a better description of what it is.

XDA-Bam committed:
Make defaults for junction angle indicator less aggressive

Default to more conservative values for decay and threshold -> less coasting.

XDA-Bam committed:
Angle indicator small cleanup

Move definition of previous_-vars closer to source

XDA-Bam committed:
Update angle indicator for 0° junctions

Make sure that d_theta_indicator and all previous_-vars are updated correctly for (close to) 0° junctions.

XDA-Bam commented:
Rebased to current HEAD after changes to JD.

thinkyhead commented:
It looks good, but of course the extra calculation is something to weigh. If the benefit is noticeable, then it would be worthwhile to make it an experimental option and see if others gain similar improvements.

I think it's a good idea to look at the number of degrees changed in inverse proportion to the distance moved, but the algorithm to do this efficiently is elusive.

XDA-Bam committed:
Adapt indicator to new acos() approximation

Using the LUT acos()-approach requires slightly different values for d_theta_threshold to work well.

XDA-Bam commented:
According to my calculations, it does work very well for our problematic test arc. I ran the early version of the code on my machine and it felt smooth. Also ran a couple of other prints and couldn't detect any apparent problem. Nonetheless, this may lead to more apparent drops in speed if the threshold is triggered after a longer coasting phase.

Default speed limitation profile with new LUT approach for `acos()` (15 elements), JD 0.1, acceleration 1000 mm/s²:
![Filter_none_speedtest_vmax_rev4](https://user-images.githubusercontent.com/1209896/81166481-f913a480-8f93-11ea-910b-31cdde9360df.png)

Speed limitation profile using d_theta_indicator:
![Filter_AGI_speedtest_vmax_rev4](https://user-images.githubusercontent.com/1209896/81166234-ad60fb00-8f93-11ea-8a18-cb8127d69169.png)

But I see the point that the added workload could be problematic on slower processors. I would encourage people to test this. If it works well as is, we can still think about increasing the effiency. If it doesn't work as expected, we would have to rethink the whole approach.

qwewer0 commented:
Should I do anything else then to use planner.cpp and planner.h? Because the test gcode feels the same.

Even if it's nothing, but I get these warnings when compiling it:
```
planner.cpp:
'limit_sqr' may be used uninitialized in this function [-Wmaybe-uninitialized] [2423, 24]
'junction_theta' may be used uninitialized in this function [-Wmaybe-uninitialized] [2424, 29]
```

XDA-Bam commented:
Just those two files, yes. The errors pop up, because I didn't initialize those two vars cleanly. It should work nonetheless. I'll fix the intialization later.

Strange that it feels the same for you. For me, it fixed the problem even with the old `acos()` approximation from 2018. Going to retest this.

qwewer0 commented:
> Could you change the following line in planner.cpp
> `Planner::d_theta_threshold = RADIANS(0.15f); // (rads/mm) Threshold for d_theta_indicator, resulting in recalculation of limit_sqr`
> to
> `Planner::d_theta_threshold = RADIANS(5.0f); // (rads/mm) Threshold for d_theta_indicator, resulting in recalculation of limit_sqr`
> and check again? 🙂

I can still feel the hiccup in the test gcode.
SKR Mini E3 v1.2, 750, 500 acc, 3000 max acc, JD 0.08

XDA-Bam commented:
I tested it today on my Ender 3 Pro (stock Melzi board, ATMega 1284p) and can confirm, that it currently does not work as expected. Neither do any of the other patches right now - alltogether, very odd. So thanks for testing, @qwewer0 !

XDA-Bam committed:
Fix angle indicator + some cleanup

Convinced the angle indicator to actually do what it's supposed to (in the test case). Changed speed limit logic, so it doesn't simply pick MAX(x,y). Instead, it keeps the first sensible value for now.

XDA-Bam commented:
OK, I've fixed it for the test case. The [test arc](https://github.com/MarlinFirmware/Marlin/issues/17342#issuecomment-613591471) runs smooth for me now, using this PR. Did not test any prints, yet.

I also found a teensy tiny error in my MATLAB vector maths, causing the estimate for `d_theta_threshold` to be about a factor of 1000 away from the correct value. Speed limit profiles now look quite different, too.

**Default bugfix-2.0.x with 15 element LUT (10e875cd7):**
![Filter_none_speedtest_vmax_rev4](https://user-images.githubusercontent.com/1209896/81432892-fbc4f400-9163-11ea-91fa-20db78fc4e57.png)

**Same branch and commit + this PR:**
![Filter_AGI_speedtest_vmax_rev4](https://user-images.githubusercontent.com/1209896/81435598-6ed06980-9168-11ea-82cd-744dd63a0245.png)

Feel free to give this a go and test it thoroughly. There could be rare cases, in which the speed limit gets "clamped" at some very low or high value. I tried to circumvent this using the `below_thresh_cnt`, but it might still happen in practice.

qwewer0 commented:
Tried it on our test gcode, and it is as good as CJ if not better.

qwewer0 commented:
Cranked the speed up to 200% with 1000 mm/s^2 acc, and it is still looks good on the small segments, but the end of the gcode gets real slow and jittery.

XDA-Bam commented:
Interesting. Will check tomorrow. By "end", you mean last part of arc, the long move back to the beginning of the arc, or the X-homing after the test?

qwewer0 commented:
> Interesting. Will check tomorrow. By "end", you mean last part of arc, the long move back to the beginning of the arc, or the X-homing after the test?

Right before the X-homing after the test.

XDA-Bam commented:
I tested you 2x GCode, the same GCode with 3000 mm/min and also 9000 mm/min. I can confirm: There's always a short stutter at the end, right before the homing move. Same with increased `BLOCK_BUFFER_SIZE = 32`. However, the same stutter at the end occurs with the regular JD using bugfix-2.0.x (10e875c). So it's not introduced by this PR, I think.

As far as I can tell, it only happens on the very last movement. I don't know where it comes from, but as long as it's only the very last one, I'd say: Let's look past that for now and test the PR in other situations.

thinkyhead commented:
> it only happens on the very last movement.

G2/G3 moves, or just your garden variety G0/G1 moves?

XDA-Bam commented:
There's only G1 in the test code. So definitely G1. Last three moves are
```
G1 X126.787 Y131.786
G1 X126.149 Y131.851
G28 X; Home X at the end
```
Stutter seems to be only on the last G1move, but it's hard to tell for sure. Stutter also apppears for regular JD without this PR.

qwewer0 commented:
> Stutter seems to be only on the last G1move

For me the slowdown and the stutter start around (if not exactly) at the second half of the curve.

XDA-Bam commented:
OK, so it doesn't behave identical. Probably because you're running an STM32F103 and mine is an ATMega 1284p. Could you see if deactivating the LUT by using `//#define JD_USE_LOOKUP_TABLE` in planner.h makes a difference?

qwewer0 commented:
> OK, so it doesn't behave identical. Probably because you're running an STM32F103 and mine is an ATMega 1284p. Could you see if deactivating the LUT by using `//#define JD_USE_LOOKUP_TABLE` in planner.h makes a difference?

Did not make any difference.

XDA-Bam commented:
OK, thanks. I'll wait until the most recent problems with JD (#17920) are fixed until I hunt for what is causing this. At first sight, it looks like we're running out of CPU cycles - but I would have expected the ATMega to be the first in line for that 😄

AnHardt commented:
As far as i can remember the ARMs we use can do CLZ(), in 32 and 64 bit, as a single assembler instruction. They should be much faster than the AVRs.
To exclude a 16 bit implementation error of GCC for that, one could try the 32bit version (instead of 16) of our lookup code.

XDA-Bam committed:
Fix definition of junction_theta in angle indicator

Since junction theta now is rads(180°)-theta, code had to be adapted. Also added some comments for clarity and removed a couple of unnecessary re-initializations.

XDA-Bam commented:
@AnHardt Could you provide the code to test this? My C++ is quite basic.

Also concerning the AVR vs. ARM thematic: How is `pgm_read_float()` resolved by the compiler for ARM-powered boards? Could this be problematic?

XDA-Bam committed:
Small optimizations to angle indicator

Moved definition of limit_sqr and junction_theta into if()-condition to avoid unnecessary intializiation. Also moved definition of previous_*-vars into same condition, to avoid updating them with zero/incorrect values.

thinkyhead commented:
> How is pgm_read_float() resolved by the compiler for ARM-powered boards?

For example…

### `$HOME/.platformio/packages/framework-arduinoststm32-maple/STM32F1/cores/maple/avr/pgmspace.h`

```cpp
#define pgm_read_float(addr) (*(const float *)(addr))
```

thinkyhead commented:
If you've got Trinamic steppers, please test with and without `SQUARE_WAVE_STEPPING` to see if there's any mitigation at the low level from this interesting feature.

XDA-Bam commented:
Needs small rebase to work with current bugfix branch. Closing because this currently does not seem to be the preferred way to fix the issue.
